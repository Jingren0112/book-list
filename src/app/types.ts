export interface IBook {
    book_author: string[],
    book_pages: number,
    book_publication_city: string,
    book_publication_country: string,
    book_publication_year: number,
    book_title: string,
    id: number
}

export enum loadingStatus {
    loading = 'loading',
    success = 'success',
    failed = 'failed'
}