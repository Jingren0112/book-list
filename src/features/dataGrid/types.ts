export enum book {
    author = 'book_author',
    pages = 'book_pages',
    publication_city = 'book_publication_city',
    publication_country = 'book_publication_country',
    publication_year = 'book_publication_year',
    title = 'book_title',
    id = 'id'
}

export const bookListColumns = [{ field: book.id },
{ field: book.title },
{ field: book.author },
{ field: book.publication_city },
{ field: book.publication_country },
{ field: book.publication_year },
{ field: book.pages }]