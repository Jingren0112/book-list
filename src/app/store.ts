import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import bookListSlice from '../features/bookList/bookListSlice';

export const store = configureStore({
  reducer: {
    bookList: bookListSlice,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
