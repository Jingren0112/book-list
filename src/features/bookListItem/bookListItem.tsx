import React from 'react';
import { isNil } from 'lodash';
import { IBook } from '../../app/types';
import { CardContent, Divider, ListItem, ListItemText, Typography } from '@mui/material';

interface IBookListItemProps {
    book: IBook | null;
}

export class BookListItem extends React.Component<IBookListItemProps> {
    render() {
        const book = this.props.book;
        return (
            <>
                <ListItem>
                    { !isNil(book) ?
                        <ListItemText disableTypography primary={ <><h1>{ `Book No: ${book.id}` }</h1> </> } secondary={
                            <CardContent component='div' sx={ { maxWidth: "40rem" } } >
                                <h2>Title: </h2>
                                <Typography component={ 'span' } >
                                    { book.book_title }
                                </Typography>
                                <h2>Author:</h2>
                                <Typography component={ 'span' } >
                                    { book.book_author }
                                </Typography>
                                <h2>Number of page:</h2>
                                <Typography component={ 'span' } >
                                    { book.book_pages }
                                </Typography>
                                <h2>Publication City:</h2>
                                <Typography component={ 'span' } >
                                    { book.book_publication_city }
                                </Typography>
                                <h2>Publication Country:</h2>
                                <Typography component={ 'span' }>
                                    { book.book_publication_country }
                                </Typography>
                                <h2> Publiccation Year: </h2>
                                <Typography component={ 'span' }>
                                    { book.book_publication_year }
                                </Typography>
                            </CardContent>
                        }
                        /> : <></>
                    }
                </ListItem>
                <Divider component="li" />
            </>
        )
    }
}

