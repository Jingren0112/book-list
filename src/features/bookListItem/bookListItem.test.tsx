import { render } from '@testing-library/react';
import { BookListItem } from './bookListItem';
import { IBook } from '../../app/types';
import { Provider } from 'react-redux';
import { store } from '../../app/store';

// Mock router
const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom') as any,
    useNavigate: () => mockedUsedNavigate,
}));

describe('bookListItem it', () => {
    it(' renders correctly without book data', () => {
        // arrange

        // act
        const tree = render(<Provider store={ store }><BookListItem book={ null } /></Provider>)
        // assert
        expect(tree).toMatchSnapshot();
    }), it(' renders correctly with book data', () => {
        // arrange
        const mockBook: IBook = {
            id: 1,
            book_title: 'test',
            book_author: ['test'],
            book_pages: 1,
            book_publication_city: 'test',
            book_publication_country: 'test',
            book_publication_year: 1
        }
        // act
        const tree = render(<Provider store={ store }><BookListItem book={ mockBook } /></Provider>)
        // assert
        expect(tree).toMatchSnapshot();
    })
})