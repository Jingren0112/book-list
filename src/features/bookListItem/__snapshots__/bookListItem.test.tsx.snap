// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bookListItem it  renders correctly with book data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      >
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <h1>
            Book No: 1
          </h1>
           
          <div
            class="MuiCardContent-root css-b7g7u3-MuiCardContent-root"
          >
            <h2>
              Title: 
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              test
            </span>
            <h2>
              Author:
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              test
            </span>
            <h2>
              Number of page:
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              1
            </span>
            <h2>
              Publication City:
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              test
            </span>
            <h2>
              Publication Country:
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              test
            </span>
            <h2>
               Publiccation Year: 
            </h2>
            <span
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              1
            </span>
          </div>
        </div>
      </li>
      <li
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        role="separator"
      />
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <h1>
          Book No: 1
        </h1>
         
        <div
          class="MuiCardContent-root css-b7g7u3-MuiCardContent-root"
        >
          <h2>
            Title: 
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            test
          </span>
          <h2>
            Author:
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            test
          </span>
          <h2>
            Number of page:
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            1
          </span>
          <h2>
            Publication City:
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            test
          </span>
          <h2>
            Publication Country:
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            test
          </span>
          <h2>
             Publiccation Year: 
          </h2>
          <span
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            1
          </span>
        </div>
      </div>
    </li>
    <li
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      role="separator"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`bookListItem it  renders correctly without book data 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      />
      <li
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
        role="separator"
      />
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    />
    <li
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      role="separator"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
